File: js/lite/src/lite/announce.ts
import * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Message from "./message.ts";
import { Version } from "./version.ts";

const MAX_HOPS = 32;

export class Announce {
	suffix: Path.Valid;
	active: boolean;

	/// Ordered origin path. Draft03 populates with 0n (UNKNOWN) entries; Draft04+ uses real IDs.
	hops: bigint[];

	constructor(props: { suffix: Path.Valid; active: boolean; hops?: bigint[] }) {
		this.suffix = props.suffix;
		this.active = props.active;
		this.hops = props.hops ?? [];
	}

	async #encode(w: Writer, version: Version) {
		await w.bool(this.active);
		await w.string(this.suffix);

		switch (version) {
			case Version.DRAFT_03:
				if (this.hops.length > MAX_HOPS) {
					throw new Error(`hop count ${this.hops.length} exceeds maximum of ${MAX_HOPS}`);
				}
				await w.u53(this.hops.length);
				break;
			case Version.DRAFT_01:
			case Version.DRAFT_02:
				break;
			default:
				if (this.hops.length > MAX_HOPS) {
					throw new Error(`hop count ${this.hops.length} exceeds maximum of ${MAX_HOPS}`);
				}
				await w.u53(this.hops.length);
				for (const hop of this.hops) {
					await w.u62(hop);
				}
				break;
		}
	}

	static async #decode(r: Reader, version: Version): Promise<Announce> {
		const active = await r.bool();
		const suffix = Path.from(await r.string());

		const hops: bigint[] = [];
		switch (version) {
			case Version.DRAFT_03: {
				// Read count but don't know actual IDs; use 0 as unknown placeholder.
				const count = await r.u53();
				if (count > MAX_HOPS) {
					throw new Error(`hop count ${count} exceeds maximum of ${MAX_HOPS}`);
				}
				for (let i = 0; i < count; i++) {
					hops.push(0n);
				}
				break;
			}
			case Version.DRAFT_01:
			case Version.DRAFT_02:
				break;
			default: {
				const count = await r.u53();
				if (count > MAX_HOPS) {
					throw new Error(`hop count ${count} exceeds maximum of ${MAX_HOPS}`);
				}
				for (let i = 0; i < count; i++) {
					hops.push(await r.u62());
				}
				break;
			}
		}

		return new Announce({ suffix, active, hops });
	}

	async encode(w: Writer, version: Version): Promise<void> {
		return Message.encode(w, (w) => this.#encode(w, version));
	}

	static async decode(r: Reader, version: Version): Promise<Announce> {
		return Message.decode(r, (r) => Announce.#decode(r, version));
	}

	static async decodeMaybe(r: Reader, version: Version): Promise<Announce | undefined> {
		return Message.decodeMaybe(r, (r) => Announce.#decode(r, version));
	}
}

export class AnnounceInterest {
	prefix: Path.Valid;

	/// Filter out announces whose hops contain this hop ID. 0n means no filtering.
	excludeHop: bigint;

	constructor(props: { prefix: Path.Valid; excludeHop?: bigint }) {
		this.prefix = props.prefix;
		this.excludeHop = props.excludeHop ?? 0n;
	}

	async #encode(w: Writer, version: Version) {
		await w.string(this.prefix);

		switch (version) {
			case Version.DRAFT_01:
			case Version.DRAFT_02:
			case Version.DRAFT_03:
				break;
			default:
				await w.u62(this.excludeHop);
				break;
		}
	}

	static async #decode(r: Reader, version: Version): Promise<AnnounceInterest> {
		const prefix = Path.from(await r.string());

		let excludeHop = 0n;
		switch (version) {
			case Version.DRAFT_01:
			case Version.DRAFT_02:
			case Version.DRAFT_03:
				break;
			default:
				excludeHop = await r.u62();
				break;
		}

		return new AnnounceInterest({ prefix, excludeHop });
	}

	async encode(w: Writer, version: Version): Promise<void> {
		return Message.encode(w, (w) => this.#encode(w, version));
	}

	static async decode(r: Reader, version: Version): Promise<AnnounceInterest> {
		return Message.decode(r, (r) => AnnounceInterest.#decode(r, version));
	}
}

/// Sent after setup to communicate the initially announced paths.
///
/// Used by Draft01/Draft02 only. Draft03+ uses individual Announce messages instead.
export class AnnounceInit {
	suffixes: Path.Valid[];

	constructor(paths: Path.Valid[]) {
		this.suffixes = paths;
	}

	static #guard(version: Version) {
		switch (version) {
			case Version.DRAFT_01:
			case Version.DRAFT_02:
				break;
			default:
				throw new Error("announce init not supported for this version");
		}
	}

	async #encode(w: Writer) {
		await w.u53(this.suffixes.length);
		for (const path of this.suffixes) {
			await w.string(path);
		}
	}

	static async #decode(r: Reader): Promise<AnnounceInit> {
		const count = await r.u53();
		const suffixes: Path.Valid[] = [];
		for (let i = 0; i < count; i++) {
			suffixes.push(Path.from(await r.string()));
		}
		return new AnnounceInit(suffixes);
	}

	async encode(w: Writer, version: Version): Promise<void> {
		AnnounceInit.#guard(version);
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader, version: Version): Promise<AnnounceInit> {
		AnnounceInit.#guard(version);
		return Message.decode(r, AnnounceInit.#decode);
	}
}

File: rs/moq-lite/src/lite/announce.rs
use num_enum::{IntoPrimitive, TryFromPrimitive};

use crate::{OriginId, Path, coding::*};

use super::{Message, Version};

impl Decode<Version> for OriginId {
	fn decode<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let value = u64::decode(r, version)?;
		Self::try_from(value).map_err(|_| DecodeError::InvalidValue)
	}
}

impl Encode<Version> for OriginId {
	fn encode<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		self.into_inner().encode(w, version)
	}
}

/// Sent by the publisher to announce the availability of a track.
/// The payload contains the contents of the wildcard.
#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum Announce<'a> {
	Active {
		#[cfg_attr(feature = "serde", serde(borrow))]
		suffix: Path<'a>,
		hops: Vec<OriginId>,
	},
	Ended {
		#[cfg_attr(feature = "serde", serde(borrow))]
		suffix: Path<'a>,
		hops: Vec<OriginId>,
	},
}

impl Message for Announce<'_> {
	fn decode_msg<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let status = AnnounceStatus::decode(r, version)?;
		let suffix = Path::decode(r, version)?;
		let hops = match version {
			Version::Lite01 | Version::Lite02 => Vec::new(),
			Version::Lite03 => {
				// Lite03 sends a single varint count; we don't know the actual IDs.
				let count = u64::decode(r, version)? as usize;
				if count > 32 {
					return Err(DecodeError::InvalidValue);
				}
				vec![OriginId::UNKNOWN; count]
			}
			_ => {
				// Lite04+: count followed by that many OriginId varints.
				let count = u64::decode(r, version)? as usize;
				if count > 32 {
					return Err(DecodeError::InvalidValue);
				}
				let mut ids = Vec::with_capacity(count);
				for _ in 0..count {
					ids.push(OriginId::decode(r, version)?);
				}
				ids
			}
		};

		Ok(match status {
			AnnounceStatus::Active => Self::Active { suffix, hops },
			AnnounceStatus::Ended => Self::Ended { suffix, hops },
		})
	}

	fn encode_msg<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		let (status, suffix, hops) = match self {
			Self::Active { suffix, hops } => (AnnounceStatus::Active, suffix, hops),
			Self::Ended { suffix, hops } => (AnnounceStatus::Ended, suffix, hops),
		};

		status.encode(w, version)?;
		suffix.encode(w, version)?;

		match version {
			Version::Lite01 | Version::Lite02 => {}
			Version::Lite03 => {
				if hops.len() > 32 {
					return Err(EncodeError::TooMany);
				}
				(hops.len() as u64).encode(w, version)?;
			}
			_ => {
				if hops.len() > 32 {
					return Err(EncodeError::TooMany);
				}
				(hops.len() as u64).encode(w, version)?;
				for id in hops {
					id.encode(w, version)?;
				}
			}
		}

		Ok(())
	}
}

/// Sent by the subscriber to request ANNOUNCE messages.
#[derive(Clone, Debug)]
pub struct AnnouncePlease<'a> {
	// Request tracks with this prefix.
	pub prefix: Path<'a>,

	/// Skip announces whose hops list contains this hop ID.
	/// Used to avoid loops in the relay cluster.
	/// Defaults to 0 (unknown), which means no filtering.
	pub exclude_hop: OriginId,
}

impl Message for AnnouncePlease<'_> {
	fn decode_msg<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let prefix = Path::decode(r, version)?;
		let exclude_hop = match version {
			Version::Lite01 | Version::Lite02 | Version::Lite03 => OriginId::UNKNOWN,
			_ => {
				let value = u64::decode(r, version)?;
				if value == 0 {
					OriginId::UNKNOWN
				} else {
					OriginId::try_from(value).map_err(|_| DecodeError::InvalidValue)?
				}
			}
		};
		Ok(Self { prefix, exclude_hop })
	}

	fn encode_msg<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		self.prefix.encode(w, version)?;
		match version {
			Version::Lite01 | Version::Lite02 | Version::Lite03 => {}
			_ => {
				self.exclude_hop.into_inner().encode(w, version)?;
			}
		}

		Ok(())
	}
}

/// Send by the publisher, used to determine the message that follows.
#[derive(Clone, Copy, Debug, IntoPrimitive, TryFromPrimitive)]
#[repr(u8)]
enum AnnounceStatus {
	Ended = 0,
	Active = 1,
}

impl Decode<Version> for AnnounceStatus {
	fn decode<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let status = u8::decode(r, version)?;
		status.try_into().map_err(|_| DecodeError::InvalidValue)
	}
}

impl Encode<Version> for AnnounceStatus {
	fn encode<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		(*self as u8).encode(w, version)
	}
}

/// Sent after setup to communicate the initially announced paths.
///
/// Used by Draft01/Draft02 only. Draft03 uses individual Announce messages instead.
#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct AnnounceInit<'a> {
	/// List of currently active broadcasts, encoded as suffixes to be combined with the prefix.
	#[cfg_attr(feature = "serde", serde(borrow))]
	pub suffixes: Vec<Path<'a>>,
}

impl Message for AnnounceInit<'_> {
	fn decode_msg<R: bytes::Buf>(r: &mut R, version: Version) -> Result<Self, DecodeError> {
		match version {
			Version::Lite01 | Version::Lite02 => {}
			_ => {
				return Err(DecodeError::Version);
			}
		}

		let count = u64::decode(r, version)?;

		// Don't allocate more than 1024 elements upfront
		let mut paths = Vec::with_capacity(count.min(1024) as usize);

		for _ in 0..count {
			paths.push(Path::decode(r, version)?);
		}

		Ok(Self { suffixes: paths })
	}

	fn encode_msg<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		match version {
			Version::Lite01 | Version::Lite02 => {}
			_ => {
				return Err(EncodeError::Version);
			}
		}

		(self.suffixes.len() as u64).encode(w, version)?;
		for path in &self.suffixes {
			path.encode(w, version)?;
		}

		Ok(())
	}
}
